File: src/2593.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class Solution {
   public:
    long long findScore(vector<int>& nums) {
        long long retVal = 0;

        int numsSize = nums.size();

        auto cmp = [](pair<int, int>& pair1, pair<int, int>& pair2) {
            if (pair1.first != pair2.first) {
                return pair1.first > pair2.first;
            }
            return pair1.second > pair2.second;
        };
        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(cmp)> heap(cmp);
        for (int i = 0; i < numsSize; ++i) {
            heap.push(make_pair(nums[i], i));
        }

        vector<bool> marked(numsSize);
        while (heap.empty() == false) {
            pair<int, int> element = heap.top();
            heap.pop();
            int num = element.first;
            int index = element.second;

            if (marked[index] == true) {
                continue;
            }

            retVal += num;

            marked[index] = true;
            // mark adjacent elements if they exist
            if (index - 1 >= 0) {
                marked[index - 1] = true;
            }
            if (index + 1 < numsSize) {
                marked[index + 1] = true;
            }
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<int> nums;
    };
    vector<subject> testData{{{2, 1, 3, 4, 5, 2}}, {{2, 3, 5, 1, 3, 2}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: nums = [2,1,3,4,5,2]
     *  Output: 7
     *
     *  Input: nums = [2,3,5,1,3,2]
     *  Output: 5
     */

    Solution cSolution;
    long long answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: nums = [";
        for (long unsigned int j = 0; j < testData[i].nums.size(); ++j) {
            cout << (j == 0 ? "" : ",") << testData[i].nums[j];
        }
        cout << "]\n";

        answer = cSolution.findScore(testData[i].nums);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/2593.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def findScore(self, nums):
        retVal = 0

        numsSize = len(nums)

        heap = []
        for index, num in enumerate(nums):
            heappush(heap, (num, index))

        marked = [False] * numsSize
        while heap:
            num, index = heappop(heap)

            if marked[index] == True:
                continue

            retVal += num
            marked[index] = True
            # mark adjacent elements if they exist
            if index - 1 >= 0:
                marked[index - 1] = True
            if index + 1 < numsSize:
                marked[index + 1] = True

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums in [[2, 1, 3, 4, 5, 2], [2, 3, 5, 1, 3, 2]]:
            # /* Example
            #  *  Input: nums = [2,1,3,4,5,2]
            #  *  Output: 7
            #  *
            #  *  Input: nums = [2,3,5,1,3,2]
            #  *  Output: 5
            #  */
            logging.info("Input: nums = %s", nums)

            retVal = pSolution.findScore(nums)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
